L1Trigger/L1THGCal/plugins/ntuples/HGCalTriggerNtupleGenTau.cc

```diff
@@ -0,0 +1,374 @@
+    {
+        isEle=true;
+    }
+    return isEle;
+}
+
+bool HGCalTriggerNtupleGenTau::isMuon( const reco::GenParticle& candidate )
+{
+    bool isMu=false;
+    if( fabs(candidate.pdgId()) == 13 && candidate.isDirectPromptTauDecayProductFinalState() && candidate.isLastCopy() )
+    {
+        isMu=true;
+    }
+    return isMu;
+}
+
+bool HGCalTriggerNtupleGenTau::isNeutralPion( const reco::GenParticle& candidate )
+{
+    bool isPiZero=false;
+    if( fabs(candidate.pdgId()) == 111 && candidate.status()==2 && candidate.statusFlags().isTauDecayProduct()
+        && !candidate.isDirectPromptTauDecayProductFinalState() )
+    {
+        isPiZero=true;
+    }
+    return isPiZero;
+}
+
+bool HGCalTriggerNtupleGenTau::isGamma( const reco::GenParticle& candidate )
+{
+    bool isGammaFromPiZero=false;
+    if( fabs(candidate.pdgId()) == 22 && candidate.status()==1 && candidate.statusFlags().isTauDecayProduct() 
+        && !candidate.isDirectPromptTauDecayProductFinalState() && candidate.isLastCopy() )
+    {
+        isGammaFromPiZero=true;
+    }
+    return isGammaFromPiZero;
+}
+
+bool HGCalTriggerNtupleGenTau::isIntermediateResonance( const reco::GenParticle& candidate )
+{
+    bool isResonance=false;
+    if( ( fabs(candidate.pdgId()) == 213 || fabs(candidate.pdgId()) == 20213 || fabs(candidate.pdgId()) == 24 )
+        && candidate.isDirectPromptTauDecayProductFinalState() && candidate.status() == 2  )
+    {
+        isResonance=true;
+    }
+    return isResonance;
+}
+
+void
+HGCalTriggerNtupleGenTau::
+fill(const edm::Event& e, const edm::EventSetup& es)
+{
+    edm::Handle<reco::GenParticleCollection> gen_particles_h;
+    e.getByToken(gen_token_, gen_particles_h);
+    const reco::GenParticleCollection& gen_particles = *gen_particles_h;
+
+    clear();
+    
+    for(const auto& particle : gen_particles)
+    {
+        
+        /* select good taus */
+        if(fabs(particle.pdgId())==15 && particle.status()==2){
+
+            LorentzVector tau_p4 = particle.p4();
+            LorentzVector tau_p4vis(0.,0.,0.,0.);
+            gen_tau_pt_.emplace_back(tau_p4.Pt());
+            gen_tau_eta_.emplace_back(tau_p4.Eta());
+            gen_tau_phi_.emplace_back(tau_p4.Phi());
+            gen_tau_energy_.emplace_back(tau_p4.E());
+            gen_tau_mass_.emplace_back(tau_p4.M());
+
+            int n_pi=0;
+            int n_piZero=0;
+            int n_gamma=0;
+            int n_ele=0;
+            int n_mu=0;
+
+            std::vector<float> tau_products_pt;
+            std::vector<float> tau_products_eta;
+            std::vector<float> tau_products_phi;
+            std::vector<float> tau_products_energy;
+            std::vector<float> tau_products_mass;
+            std::vector< int > tau_products_id;
+            
+            /* loop over tau daughters */
+            const edm::RefVector<std::vector<reco::GenParticle> >& daughters = particle.daughterRefVector();
```

Comment:
Use `GenParticleRefVector` instead of `edm::RefVector<std::vector<reco::GenParticle> >`